### PR TITLE
perf(model): Add xformer and pytorch in LLaMA

### LINKED ISSUE #72: [Super AI] [Good First Issue] Convert LLaMa to XFormer Attention
Description:
Although the Official Implementation of [LLaMa Model](https://github.com/facebookresearch/llama) mentioned they used [XFormer Attention](https://github.com/facebookresearch/xformers) in their paper to have efficient memory consumption when training, did not provide an example of doing so in their [inference code](https://github.com/facebookresearch/llama/blob/main/llama/model.py). In this task, we want to convert the original code from LLaMA to support Xformer Attetnion by doing

1. Download the model https://github.com/facebookresearch/llama/pull/73 and try to run inference using their official implementation code. (6.7B Version is enough)
2. Convert to both:
- Pytorch 2.0 Attention: https://pytorch.org/docs/stable/generated/torch.nn.functional.scaled_dot_product_attention.html
- Xformer Attention 
3. Try to run inference again and get the same result (logits)

Deliverables:
- Convert/Inference Code Script
- PR
- Unit Test to make sure converted model have similar output to the original one

### PR DESCRIPTION
**Pull request by boss-chanon**
add xformer and pytorch in LLaMA for `generation.py` and `tokenizer.py` is copied from [llama from facebook](https://github.com/facebookresearch/llama) and `model.py` change from use fairscale to use torch and modify to can use torch attention and xformer attention

close #72

### CODE REVIEW

**Review by boat1603 — COMMENTED** [review]
Could you explain me more about how different of `generation.py`, `model.py`, and `tokenizer.py` from [llama from facebook](https://github.com/facebookresearch/llama). This PR is too long. Your explanation can help us to review your code easier.

**Comment by new5558** [comment]
Note: Will only use LLaMa for experiment and convert code to RedLLaMa later.

File: src/model/openthaigpt_pretraining_model/llama/generation.py [hunk 1]
```diff
@@ -0,0 +1,79 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates.
+# This software may be used and distributed according to the terms of the GNU General Public License version 3.
+
+# This code copy from https://github.com/facebookresearch/llama/blob/main/llama/generation.py
```
[the remark this answers is not in the document]
Comment (boss-chanon) [reply src/model/openthaigpt_pretraining_model/llama/generation.py:4]: `generation.py` and `tokenizer.py` is copied from [llama from facebook](https://github.com/facebookresearch/llama). and `model.py` modify to can use xformer and pytorch

File: src/model/openthaigpt_pretraining_model/llama/model.py [hunk 2]
```diff
@@ -1,14 +1,244 @@
...
+            output = torch.matmul(scores, values)  # (bs, n_local_heads, slen, head_dim)
+        
+        output = output.transpose(
+            1, 2
+        ).contiguous().view(bsz, seqlen, -1)
+
+        return self.wo(output)
+
+
+class FeedForward(nn.Module):
```
Comment (boat1603) [inline src/model/openthaigpt_pretraining_model/llama/model.py:174]: This `FeedForward` model you chenged only `self.w1`, `self.w2`, and `self.w3` from `ColumnParallelLinear` and `RowParallelLinear` to `nn.Linear` but no changed in forward function. I suggest you to inherit `FeedForward` from original implementation.

This is example of my code.
```python
from llama.model import FeedForward as OriginalLlamaFeedForward

class FeedForward(OriginalLlamaFeedForward):
    def __init__(
        self,
        dim: int,
        hidden_dim: int,
        multiple_of: int,
    ):
        super().__init__(dim, hidden_dim, multiple_of)
        self.w1 = nn.Linear(
            dim, hidden_dim, bias=False
        )
        self.w2 = nn.Linear(
            hidden_dim, dim, bias=False
        )
        self.w3 = nn.Linear(
            dim, hidden_dim, bias=False
        )
```

File: src/model/openthaigpt_pretraining_model/llama/model.py [hunk 3]
```diff
@@ -1,14 +1,244 @@
...
+        )
+        self.w3 = nn.Linear(
+            dim, hidden_dim, bias=False
+        )
+
+    def forward(self, x):
+        return self.w2(F.silu(self.w1(x)) * self.w3(x))
+
+
+class TransformerBlock(nn.Module):
```
Comment (boat1603) [inline src/model/openthaigpt_pretraining_model/llama/model.py:199]: Ditto

File: src/model/openthaigpt_pretraining_model/llama/model.py [hunk 4]
```diff
@@ -1,14 +1,244 @@
...
+        self.attention_norm = RMSNorm(args.dim, eps=args.norm_eps)
+        self.ffn_norm = RMSNorm(args.dim, eps=args.norm_eps)
+
+    def forward(self, x: torch.Tensor, start_pos: int, freqs_cis: torch.Tensor, mask: Optional[torch.Tensor]):
+        h = x + self.attention.forward(self.attention_norm(x), start_pos, freqs_cis, mask)
+        out = h + self.feed_forward.forward(self.ffn_norm(h))
+        return out
+
+
+class Transformer(nn.Module):
```
Comment (boat1603) [inline src/model/openthaigpt_pretraining_model/llama/model.py:219]: Ditto

**Comment by ArthurMinovsky** [comment]
You need to wait to pull the request checker change your code format first.  And then, You can commit again and waiting reviewer to approve your code

File: src/model/openthaigpt_pretraining_model/llama/generation.py [hunk 5]
```diff
@@ -0,0 +1,79 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates.
```
[the remark this answers is not in the document]
Comment (boss-chanon) [reply src/model/openthaigpt_pretraining_model/llama/generation.py]: done

File: src/model/openthaigpt_pretraining_model/llama/tokenizer.py [hunk 6]
```diff
@@ -0,0 +1,42 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates.
```
[the remark this answers is not in the document]
Comment (boss-chanon) [reply src/model/openthaigpt_pretraining_model/llama/tokenizer.py]: done

File: src/model/openthaigpt_pretraining_model/llama/model.py (hunk 2 above)
Comment (boss-chanon) [reply src/model/openthaigpt_pretraining_model/llama/model.py]: If use this way we will set up fairscale that useless

Comment (boat1603) [reply src/model/openthaigpt_pretraining_model/llama/model.py]: How?

Comment (boat1603) [reply src/model/openthaigpt_pretraining_model/llama/model.py]: How long setup time?

Comment (boss-chanon) [reply src/model/openthaigpt_pretraining_model/llama/model.py]: I have never setup and fairscale @new5558 called will don't use in this project

File: src/model/openthaigpt_pretraining_model/llama/model.py [hunk 7]
```diff
@@ -1,14 +1,208 @@
...
+            (args.max_batch_size, args.max_seq_len, self.n_local_heads, self.head_dim)
+        )
+        self.cache_v = torch.zeros(
+            (args.max_batch_size, args.max_seq_len, self.n_local_heads, self.head_dim)
+        )
+
+        if not (
+            args.attention_mode == "origin"
+            or args.attention_mode == "pytorch"
+            or args.attention_mode == "xformer"
```
[the remark this answers is not in the document]
Comment (boss-chanon) [reply src/model/openthaigpt_pretraining_model/llama/model.py:76]: done

File: src/model/openthaigpt_pretraining_model/llama/model.py [hunk 8]
```diff
@@ -1,14 +1,215 @@
...
+    n_layers: int = 8
+    n_heads: int = 8
+    vocab_size: int = -1  # defined later by tokenizer
+    multiple_of: int = 256  # make SwiGLU hidden layer size multiple of large power of 2
+    norm_eps: float = 1e-5
+
+    max_batch_size: int = 32
+    max_seq_len: int = 2048
+
+    attention_mode: str = "origin"  # pytorch, xformer
```
Comment (boat1603) [inline src/model/openthaigpt_pretraining_model/llama/model.py:34]: const

File: tests/model/llama_test.py [hunk 9]
```diff
@@ -1,21 +1,42 @@
...
+    torch_model.load_state_dict(base_model.state_dict())
+
+    for test_case in LLAMA_TEST_CASES:
+        assert torch.all(
+            torch.abs(base_model(test_case, 0) - torch_model(test_case, 0)) <= 1e-6
+        )
+
+
+def test_llama_efficient_for_xformer():
+    base_args = ModelArgs(vocab_size=VOCAB_SIZE, attention_mode="origin")
```
Comment (boat1603) [inline tests/model/llama_test.py:31]: const

File: tests/model/llama_test.py [hunk 10]
```diff
@@ -1,21 +1,42 @@
...
+    xformer_args = ModelArgs(vocab_size=VOCAB_SIZE, attention_mode="xformer")
```
Comment (boat1603) [inline tests/model/llama_test.py:32]: const

File: tests/model/llama_test.py [hunk 11]
```diff
@@ -1,21 +1,42 @@
...
+    torch.randint(0, VOCAB_SIZE, (1, 1)),
+    torch.randint(0, VOCAB_SIZE, (12, 1)),
+    torch.randint(0, VOCAB_SIZE, (1, 12)),
+    torch.randint(0, VOCAB_SIZE, (12, 12)),
+    torch.randint(0, VOCAB_SIZE, (6, 4)),
+]
+
+
+def test_llama_efficient_for_torch():
+    base_args = ModelArgs(vocab_size=VOCAB_SIZE, attention_mode="origin")
```
Comment (boat1603) [inline tests/model/llama_test.py:16]: const

File: tests/model/llama_test.py [hunk 12]
```diff
@@ -1,21 +1,42 @@
...
+    torch_args = ModelArgs(vocab_size=VOCAB_SIZE, attention_mode="pytorch")
```
Comment (boat1603) [inline tests/model/llama_test.py:17]: const

File: src/model/openthaigpt_pretraining_model/llama/model.py (hunk 8 above)
Comment (boss-chanon) [reply src/model/openthaigpt_pretraining_model/llama/model.py]: done

File: tests/model/llama_test.py (hunk 9 above)
Comment (boss-chanon) [reply tests/model/llama_test.py]: done

File: tests/model/llama_test.py (hunk 10 above)
Comment (boss-chanon) [reply tests/model/llama_test.py]: done

File: tests/model/llama_test.py (hunk 11 above)
Comment (boss-chanon) [reply tests/model/llama_test.py]: done

File: tests/model/llama_test.py (hunk 12 above)
Comment (boss-chanon) [reply tests/model/llama_test.py]: done

**Review by boat1603 — APPROVED** [review]

File: src/model/openthaigpt_pretraining_model/llama/model.py [hunk 13]
```diff
@@ -1,14 +1,215 @@
...
+from dataclasses import dataclass
+import math
+
+import torch
+from torch import nn
+import torch.nn.functional as F  # noqa: N812
+
+from torch.nn.functional import scaled_dot_product_attention as torch_attention
+from xformers.components.attention.core import (
+    scaled_dot_product_attention as xformer_attention,
```
Comment (new5558) [inline src/model/openthaigpt_pretraining_model/llama/model.py:11]: Please use this one instead: https://facebookresearch.github.io/xformers/components/ops.html

File: src/model/openthaigpt_pretraining_model/llama/model.py [hunk 14]
```diff
@@ -1,14 +1,215 @@
...
+        values = self.cache_v[:bsz, : start_pos + seqlen]
+
+        xq = xq.transpose(1, 2)
+        keys = keys.transpose(1, 2)
+        values = values.transpose(1, 2)
+
+        if self.mode == PYTORCH_ATTENTION_MODE:
+            output = torch_attention(xq, keys, values, mask)
+        elif self.mode == XFORMER_ATTENTION_MODE:
+            output = xformer_attention(xq, keys, values, mask)
```
Comment (new5558) [inline src/model/openthaigpt_pretraining_model/llama/model.py:116]: for xformer you can use 
`LowerTriangularMask ` instead of the mask variable

example:
```python
import xformers.ops as xops

# Compute regular attention
y = xops.memory_efficient_attention(q, k, v)

# With a dropout of 0.2
y = xops.memory_efficient_attention(q, k, v, p=0.2)

# Causal attention
y = xops.memory_efficient_attention(
    q, k, v,
    attn_bias=xops.LowerTriangularMask()
)
```

source: https://facebookresearch.github.io/xformers/components/ops.html#xformers.ops.fmha.attn_bias.LowerTriangularMask

File: src/model/openthaigpt_pretraining_model/llama/model.py [hunk 15]
```diff
@@ -1,14 +1,215 @@
...
+        keys = self.cache_k[:bsz, : start_pos + seqlen]
+        values = self.cache_v[:bsz, : start_pos + seqlen]
+
+        xq = xq.transpose(1, 2)
+        keys = keys.transpose(1, 2)
+        values = values.transpose(1, 2)
+
+        if self.mode == PYTORCH_ATTENTION_MODE:
+            output = torch_attention(xq, keys, values, mask)
+        elif self.mode == XFORMER_ATTENTION_MODE:
```
Comment (new5558) [inline src/model/openthaigpt_pretraining_model/llama/model.py:115]: there is another usecase of xformer attention mask **(and it should be faster!)**. It is called CasualBlockDiagonalMask 
please check if the input is a list of tensors. If yes, use `CasualBlockDiagonalMask` instead of `LowerTriangularMaskWithTensorBias`

Example of how to use `CasualBlockDiagonalMask`
```python
import torch
from xformers.ops import fmha

K = 16
dtype = torch.float16
device = "cuda"
list_x = [
    torch.randn([1, 3, 1, K], dtype=dtype, device=device),
    torch.randn([1, 6, 1, K], dtype=dtype, device=device),
    torch.randn([1, 2, 1, K], dtype=dtype, device=device),
]
attn_bias, x = fmha.BlockDiagonalMask.from_tensor_list(list_x)
linear = torch.nn.Linear(K, K * 3).to(device=device, dtype=dtype)

q, k, v = linear(x).reshape([1, -1, 1, 3, K]).unbind(-2)
out = fmha.memory_efficient_attention(q, k, v, attn_bias=attn_bias)
list_out = attn_bias.split(out)
print(list_out[0].shape)  # [1, 3, 1, K]
assert tuple(list_out[0].shape) == (1, 3, 1, K)
```
source: https://facebookresearch.github.io/xformers/components/ops.html#xformers.ops.fmha.attn_bias.BlockDiagonalMask.from_tensor_list

File: src/model/openthaigpt_pretraining_model/llama/model.py [hunk 16]
```diff
@@ -1,14 +1,215 @@
...
+
+        self.norm = RMSNorm(params.dim, eps=params.norm_eps)
+
+        self.output = nn.Linear(params.dim, params.vocab_size)
+
+        self.freqs_cis = precompute_freqs_cis(
+            self.params.dim // self.params.n_heads, self.params.max_seq_len * 2
+        )
+
+    @torch.inference_mode()
```
Comment (new5558) [inline src/model/openthaigpt_pretraining_model/llama/model.py:197]: remove this

**Review by new5558 — CHANGES_REQUESTED** [review]

File: src/model/openthaigpt_pretraining_model/llama/model.py (hunk 15 above)
Comment (boss-chanon) [reply src/model/openthaigpt_pretraining_model/llama/model.py:115]: input is only `torch.Tensor` can not be `list` of `torch.Tensor`

File: src/model/openthaigpt_pretraining_model/llama/model.py (hunk 14 above)
Comment (boss-chanon) [reply src/model/openthaigpt_pretraining_model/llama/model.py:116]: i am try to use 
`y = xops.memory_efficient_attention(
    q, k, v,
    attn_bias=xops.LowerTriangularMask()
)`
but it have more error

File: src/model/openthaigpt_pretraining_model/llama/model.py (hunk 16 above)
Comment (boss-chanon) [reply src/model/openthaigpt_pretraining_model/llama/model.py]: done

File: src/model/openthaigpt_pretraining_model/llama/model.py [hunk 17]
```diff
@@ -1,14 +1,215 @@
...
+        self.cache_v = torch.zeros(
+            (args.max_batch_size, args.max_seq_len, self.n_local_heads, self.head_dim)
+        )
+
+        if not (
+            args.attention_mode == PYTORCH_ATTENTION_MODE
+            or args.attention_mode == XFORMER_ATTENTION_MODE
+            or args.attention_mode == ORIGIN_ATTENTION_MODE
+        ):
+            raise KeyError(
```
[the remark this answers is not in the document]
Comment (boss-chanon) [reply src/model/openthaigpt_pretraining_model/llama/model.py:78]: i am already test it or you want to test it in pytest

[the remark this answers is not in the document]
Comment (new5558) [reply src/model/openthaigpt_pretraining_model/llama/model.py:78]: Yes

File: src/model/openthaigpt_pretraining_model/llama/model.py (hunk 14 above)
Comment (new5558) [reply src/model/openthaigpt_pretraining_model/llama/model.py:116]: You can use if `torch.is_cuda_available` to handle this case. What do you mean when saying "high discrepancy"

File: src/model/openthaigpt_pretraining_model/llama/model.py (hunk 17 above)
[the remark this answers is not in the document]
Comment (boss-chanon) [reply src/model/openthaigpt_pretraining_model/llama/model.py:75]: done

File: src/model/openthaigpt_pretraining_model/llama/model.py (hunk 13 above)
Comment (boss-chanon) [reply src/model/openthaigpt_pretraining_model/llama/model.py]: done

File: src/model/openthaigpt_pretraining_model/llama/model.py (hunk 14 above)
Comment (boss-chanon) [reply src/model/openthaigpt_pretraining_model/llama/model.py]: now i can fix high discrepancy and when input use `cpu` i will switch to origin attention

**Review by new5558 — APPROVED** [review]
LGTM
